### PR TITLE
Disallow CNS services with dots in the name.

### PR DESCRIPTION
**Pull request by kwilczynski**
This commit adds validation for dots (among other things) in the CNS services names (which is not
supported by the CloudAPI, but wasn't handled properly in the provider), and thus resolves and issue where Terraform would time out and terminate with an error, even thought a compute instance would be created with an appropriate CNS service configured, and it would be completely healthy.

Signed-off-by: Krzysztof Wilczynski <kw@linux.com>

### CODE REVIEW

**Comment by kwilczynski** [comment]
I need to add tests, thus work in progress.

File: triton/resource_machine.go [hunk 1]
```diff
@@ -973,6 +978,26 @@ func castToTypeList(sliceRaw interface{}) []string {
...
+// domain name intact if it does not include a known delimiter to split
+// on (e.g., given domain name is not compliant with a typical CNS domain
+// name schema, etc., although this will almost never be the case).
+func extractCNSPrefix(domain string) string {
+	// The "svc" and "inst" are an integral part of the CNS
+	// domain name and thus we try to split the domain name on
+	// these, before deferring to an old behaviour of splitting
+	// on a single to get the prefix. Otherwise, we return a
+	// whole domain name as-is intact.
+	for _, ss := range []string{"svc", "inst"} {
```
Comment (cheapRoc) [inline triton/resource_machine.go]: Isn't there a chance that users will use `svc` in a CNS service name? Why not split on `.svc.` and `.inst.`?

Comment (kwilczynski) [reply triton/resource_machine.go]: Possibly. I will fix it.

**Comment by kwilczynski** [comment]
I have renamed the subject (name) of the Pull Request, and edited the commit message to reflect what type of changes will be introduced here. I did not change the name of the branch (it's not necessary to to do it, and it would only create more commotion).

**Comment by kwilczynski** [comment]
It is becoming more apparent, that this might turn into a documentation update/fix, rather than something that can be done sensibly in the code, primarily due to Terraform's limitation:

```
 ValidateFunc is not yet supported on lists or sets.
```

**Comment by kwilczynski** [comment]
I am going to close this, since Joyent Public Cloud (JPC) is no [longer](https://docs.joyent.com/joyent-public-cloud-eol).

**Comment by siepkes** [comment]
While the JPC is gone Triton itself is open source and people are running their own private clouds with it. So I think this PR might still be useful.

**Comment by kwilczynski** [comment]
Hi @siepkes, thank you for the response!

> While the JPC is gone Triton itself is open source and people are running their own private clouds with it. So I think this PR might still be useful.

That is very true, indeed. Sadly, I do not have an Open Source installation of Trition or Smart Data Center (SDC) at my disposal, and with the closure of JPC, and also me no longer working for Joyent, there is very little I can do to work on this.

If someone is willing to provide access, then I will be happy to help more. Otherwise, without spending my own time and resources on this, it might be hard to carry-on working on supporting Triton for me.

**Comment by siepkes** [comment]
> If someone is willing to provide access, then I will be happy to help more. Otherwise, without spending my own time and resources on this, it might be hard to carry-on working on supporting Triton for me.

I totally understand that.

I might be able to provide such access however I think there is a larger issue here and that is that there isn't much movement in this Terraform provider as a whole. In the sense that I think no-one is currently shepherding this provider right now as far as I can tell (PR's are not merged)? 

Since I have an interest in this I'll raise this on the Triton mailing list. Maybe as a community we can help out in this.

**Pull request closed without merging**